source/core/shape.cc

```diff
@@ -1576,35 +1606,46 @@ RBFShape<dim>::gradient(const Point<dim> &evaluation_point,
         "of at least one RBF node. It this isn't the case, an error has been "
         "introduced in the code. ");
     }
-  for (const size_t &node_id : iterable_nodes)
+  // Here we loop on every portion (of RBF nodes located in active cells close
+  // to the evaluation point) and on every RBF node located in these active
+  // cells.
+  for (size_t portion_id = 0; portion_id < iterable_nodes.size(); portion_id++)
     {
-      // Calculation of the dr/dx
-      relative_position = (evaluation_point - rotated_nodes_positions[node_id]);
-      distance          = (relative_position).norm();
-      normalized_distance = distance / support_radii[node_id];
-      if (distance > 1e-16)
-        dr_dx_derivative = relative_position / distance;
-      else
+      for (const size_t &node_id : *(std::get<2>(iterable_nodes[portion_id])))
         {
-          // If the evaluation point overlaps with the node position, we assume
-          // that the contribution of this node is 0. This assumption can be
-          // made because radial basis functions are symmetrical. If the basis
-          // function is not differentiable at its node (e.g. linear function),
-          // this approximation will still hold since the approximated distance
-          // field is already imperfect and neighboring nodes will add their
-          // contribution at the evaluation point.
-          for (int d = 0; d < dim; d++)
-            dr_dx_derivative[d] = 0;
+          // Calculation of dr/dx
+          relative_position =
+            (evaluation_point - rotated_nodes_positions[node_id]);
+          distance            = (relative_position).norm();
+          normalized_distance = distance / support_radii[node_id];
+          if (distance > 1e-16)
```

Comment:
Not the biggest fan of using tolerances which are not based on a dimension of the problem when it's a geometry thing